src/app/interaction-model/Model.h

```diff
@@ -0,0 +1,123 @@
+namespace app {
+namespace InteractionModel {
+
+/// Represents operations against a matter-defined data model.
+///
+/// Class is SINGLE-THREADED:
+///   - operations are assumed to only be ever run in a single event-loop
+///     thread or equivalent
+///   - class is allowed to attempt to cache indexes/locations for faster
+///     lookups of things (e.g during iterations)
+class Model : public AttributeTreeIterator
+{
+public:
+    virtual ~Model() = default;
+
+    // `actions` pointers  will be guaranteed valid until Shutdown is called()
+    virtual CHIP_ERROR Startup(InteractionModelActions actions)
+    {
+        mActions = actions;
+        return CHIP_NO_ERROR;
+    }
+    virtual CHIP_ERROR Shutdown() = 0;
+
+    // During the transition phase, we expect a large subset of code to require access to
+    // event emitting, path marking and other operations
+    virtual InteractionModelActions CurrentActions() { return mActions; }
+
+    /// List reading has specific handling logic:
+    ///   `state` contains in/out data about the current list reading. MUST start with kInvalidListIndex on first call
+    ///
+    /// Return codes:
+    ///   CHIP_ERROR_MORE_LIST_DATA_AVAILABLE (NOTE: new error defined for this purpose)
+    ///      - partial data written to the destination
+    ///      - destination will contain AT LEAST one valid list entry fully serialized
+    ///      - destination will be fully valid (it will be rolled back on partial list writes)
+    ///   CHIP_IM_GLOBAL_STATUS(code):
+    ///      - error codes that are translatable in IM status codes (otherwise we expect Failure to be reported)
+    ///      - In particular, some handlers rely on special handling for:
+    ///        - `UnsupportedAccess`  - for ACL checks (e.g. wildcard expansion may choose to skip these)
+    ///      - to check for this, CHIP_ERROR provides:
+    ///        - ::IsPart(ChipError::SdkPart::kIMGlobalStatus) -> bool
+    ///        - ::GetSdkCode() -> uint8_t to translate to the actual code
+    virtual CHIP_ERROR ReadAttribute(const ReadAttributeRequest & request, ReadState & state, AttributeValueEncoder & encoder) = 0;
+
+    /// Requests a write of an attribute.
+    ///
+    /// When this is invoked, caller is expected to have already done some validations:
+    ///    - cluster `data version` has been checked for the incoming request if applicable
+    ///
+    /// List operation support:
+    ///    - the first list write will have `request.writeFlags.Has(WriteFlags::kListBegin)`
+    ///    - the last list write will have `request.writeFlags.Has(WriteFlags::kListEnd)`
+    ///    - the last list write MAY have empty data (no list items)
+    ///
+    /// When `request.writeFlags.Has(WriteFlags::kForceInternal)` the request is from an internal app update
+    /// and SHOULD bypass some internal checks (like timed enforcement, potentially read-only restrictions)
+    ///
+    /// Return codes
+    ///   CHIP_IM_GLOBAL_STATUS(code):
```

Comment:
It seems more convenient because all our macros are `VerifyOrReturnError` and not `VerifyOrReturnStatus`. The goal here should be to make the handler code easier to write and using CHIP_ERROR seems to make that easier.